### PR TITLE
Adds knapsack_pro to the stack; updates configs

### PR DESCRIPTION
**Pull request by filipefurtad0**
#### What? Why?

Adds knapsack_pro to the stack.

- created an account https://knapsackpro.com/
- set the API key under the repo's secrets
- followed the instructions on https://docs.knapsackpro.com/knapsack_pro-ruby/guide/
- changed out build.yml file for the build to run a first time (benchmarking step)

#### What should we test?

- green build
- build runs should be visible under the project's dashboard

#### Release notes

Changelog Category: echnical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by filipefurtad0** [comment]
Hum, I've set the API key under the secrets section of the repo:

image

on the build file, calling it as:
`          KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC: ${{ secrets.KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC }}`

But this is not being picked up. Any ideas?

**Comment by jibees** [comment]
I've added as a "Repository Secret", without any success ;) 

Capture d’écran 2022-10-27 à 14 37 55

**Review by mkllnk — COMMENTED** [review]
Nice work. We need to find a way to separate certain specs. Maybe with rspec tags? I couldn't do it based on directories in the past but I never tried Knapsack Pro.

File: Gemfile
Comment (filipefurtad0) [reply Gemfile:170]: Done [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/9920).

File: spec/spec_helper.rb
Comment (filipefurtad0) [reply spec/spec_helper.rb:21]: Hum, good point. Reason I've added here is because we still have plenty of specs using spec_helper.rb.

I've just checked and we use in specs contained in folders like request, controllers, helpers, jobs, lib, mailers, migrations, model, performance, queries, requests, routing, serializers, services, support, validators and view specs.

Basically, everything which does not require capybara/cuprite. I guess this could be a criteria to draw the line, and try to apply knapsack only to the "bottleneck" specs which are always system specs, I guess.

File: bin/knapsack_pro_rspec
Comment (filipefurtad0) [reply bin/knapsack_pro_rspec:9]: Thanks! This is helpful. It seems to be possible to feed a list or set criteria for inclusion/exclusion: https://knapsackpro.com/faq/question/how-to-run-a-specific-list-of-test-files-or-only-some-tests-from-test-file

I'll see if I can find out the best way to separate these.

**Comment by filipefurtad0** [comment]
The updates:
- I've added the keys on the files. So the secrets issue now looks solved (running on forked repo)

I think we might instead update this setting at the repo level -> https://github.com/organizations/openfoodfoundation/settings/actions, and enable:

image

- splitted this into different build run files and used `KNAPSACK_PRO_TEST_FILE_PATTERN` option to run only some specs (ex. only `system`, or only `controllers`)

However, the first run should create a file which should persist after the build finishes. It is now throwing the error:

```
 E, [2022-11-01T01:06:36.080582 #2899] ERROR -- : [knapsack_pro] ---------- Configuration error --------------------------------------------------
E, [2022-11-01T01:06:36.080639 #2899] ERROR -- : [knapsack_pro] You forgot to call KnapsackPro::Adapters::RSpecAdapter.bind method in your test runner configuration file. It is needed to record test files time execution. Please follow the installation guide to configure your project properly https://docs.knapsackpro.com/knapsack_pro-ruby/guide/
E, [2022-11-01T01:06:36.080654 #2899] ERROR -- : [knapsack_pro] If you already have KnapsackPro::Adapters::RSpecAdapter.bind method added and you still see this error then one of your tests must have deleted the .knapsack_pro directory from the disk accidentally. Please ensure you do not remove the .knapsack_pro directory: https://knapsackpro.com/faq/question/why-all-test-files-have-01s-time-execution-for-my-ci-build-in-user-dashboard
Error: Process completed with exit code 1.

```

File: spec/spec_helper.rb
Comment (mkllnk) [reply spec/spec_helper.rb]: I'm not sure if we are understanding each other. The spec_helper is loading the base_helper and that one is already doing the setup. So I think that we don't need to do it twice. The base_helper will do it for all.

Comment (filipefurtad0) [reply spec/spec_helper.rb]: > spec_helper is loading the base_helper

Right, and not the other way around. Thanks for pointing that out.

Knapsack_pro config was removed from spec_helper 2f8a46b9bf51aad3159df29b800ba07c347f0e8b.

**Comment by filipefurtad0** [comment]
Alright - green build :green_circle:
This is when running the build in `regular mode`, a second time.

System and controller specs have a run time under 10 minutes :eyes: 

Still to do:

- generate a separate API key for the `test-the-rest / rspec (spec/models)` build 
- try the Queue Mode -> I suspect we'll have some order dependent build failures
- we can still increase the number of parallel jobs (41 currently, 19 left)
- if all is well, this branch will need a major clean-up in git-history :broom: 
- also, we should clarify if we are eligible to use the pro -> seems to be the case but not yet done

**Comment by filipefurtad0** [comment]
Ok, some updates on this PR

1) Open source plan - Pro version - is enabled for OFN 
image

2) Build seems stable: in general and with queue mode enabled for system specs. We can see the effect queue mode has on the spec running

**Before** enabling queue mode 

image
 
**After** enabling queue mode

image

3) So this is all rather conservative implementation of the gem, but it seems quite stable. We could still think of enabling additional splitting of test cases on runs: https://knapsackpro.com/faq/question/how-to-split-slow-rspec-test-files-by-test-examples-by-individual-it

So... Maybe this is ready for review? :eyes: 

PS: I've just spotted [a failing test](https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/3462051988/attempts/1) on `engines`, but re-triggering turned green quickly, and I'm not sure this relates to knapsack

`EPROTO 139987936114624:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../deps/openssl/openssl/ssl/record/rec_layer_s3.c:1546:SSL alert number 40`

File: .github/workflows/build_knapsack_system.yml
Comment (dacook) [inline .github/workflows/build_knapsack_system.yml:86]: I'm ignorant about Knapsack Pro, but I can see a comment above that says:
> if you use Knapsack Pro Queue Mode you must set below env variable...

Is this relevant to set up? If not, perhaps it's worth updating the comment to say why. Or perhaps it will be obvious once I read more..

**Review by dacook — COMMENTED** [review]
Hi @filipefurtad0 , I took a look to see if I can review this.
It looks like most of this was reviewed already, and there are already fixes relating to some of the earlier commits.  Is this correct or is it worth going through all of it?

Also I've added a query about one of the new commits.

File: .github/workflows/build_knapsack_system.yml
Comment (filipefurtad0) [reply .github/workflows/build_knapsack_system.yml:86]: Yes, indeed - this seems relevant @dacook , thanks. As I understood it from that link, we might need it in case the build fails, so it is repeated on the same order. So far, I've only enabled queue mode for system specs - so I've set `KNAPSACK_PRO_FIXED_QUEUE_SPLIT: true`
now. It should not change the outcome of the build for now - let's see :-)

**Review by dacook — APPROVED** [review]
Under the assumption that most of it has been reviewed and resolved, the new commits look good to me!

I did notice that there's a lot of duplication in the yaml workflows now (eg Postgres config), but I don't know if there's any ways to share the config between workflows.

**Comment by jibees** [comment]
@filipefurtad0 : now `spec/support/request/stripe_helper.rb` is in conflict ;(

**Comment by filipefurtad0** [comment]
Resolved conflicts and rebased.

> I don't know if there's any ways to share the config between workflows

I agree it would be great not to repeat those lines. I'm not sure how to do this without actually sharing the config - hence the approach of having them in separate files, each with its own API key. This assures these tests are run separately and we don't have order related failures by mixing them in the setup.

One question though @jibees @dacook  :
I'm not totally sure about 9784db4ec31d138372179d6db51373aace43dd10. Perhaps it would be more defensive if we keep the env variable:
`unless ENV['DISABLE_KNAPSACK_PRO']`

and somehow keep the current `build.yml` file for now as well, in case we disable Knapsack Pro. What do you think?

**Comment by dacook** [comment]
> to control which build we run

Hi @filipefurtad0, based on what I know, I don't think it's necessary to have the condition here. The point of all your (awesome) work is so that we can use Knapsack Pro all the time. So I don't see a need to be able to easily disable/enable with a variable. Do you see a potential reason we might need to do that?

Of course, if one day we decide we no longer want to use Knapsack Pro, then we would go ahead and change the config again. We might choose to revert some of this, or simply move forward with new config.

**Comment by dacook** [comment]
> No particular reason, it just feels like a big change 😅

Fair enough, I understand how you feel! 
But as it doesn't affect production code, I reckon any issues can wait long enough for us to fix with another PR.

**Comment by dacook** [comment]
So I think no Dev or User testing required, this is ready to go.

I've updated the status, but should I also merge it? (I forget the exact process sorry, and do we have that documented somewhere?)

**Comment by jibees** [comment]
> I've updated the status, but should I also merge it? (I forget the exact process sorry, and do we have that documented somewhere?)

If you've the right to do so, and you feel comfortable with it, yes, you can merge it!

**Comment by sigmundpetersen** [comment]
Just to understand. What is stopping us from running all tests from the same workflow file?
And just split them by different jobs inside the same file?
As far as I can see, all files have the same content until reaching the `jobs` line

I did read this:
>  I agree it would be great not to repeat those lines. I'm not sure how to do this without actually sharing the config - hence the approach of having them in separate files, each with its own API key. This assures these tests are run separately and we don't have order related failures by mixing them in the setup.

but the API key is also inside each job, so could still be separated by jobs, not files?
Are there order related failures when separating by jobs only?

I trust there is a reason for doing it this way, I just would like to understand 🙂

**Comment by filipefurtad0** [comment]
Sure, @sigmundpetersen - thanks for asking this. 

> Are there order related failures when separating by jobs only?
> but the API key is also inside each job, so could still be separated by jobs, not files?

So, we would still have separate DBs - the code would still be repeated, I guess - but it would be in the same file. This should be ok, I guess - worth a try!

File: .github/workflows/build_knapsack_pro.yml
Comment (sigmundpetersen) [inline .github/workflows/build_knapsack_pro.yml:90]: I think you need separate names/tags for the separate jobs, so i.e. `rspec` -> `rspec_models` or `models`

Comment (filipefurtad0) [reply .github/workflows/build_knapsack_pro.yml]: Thank you for that tip @sigmundpetersen!! Well spotted :pray: 
The build now runs all nodes, all in one file, which is great: it's much easier to have an overview of the whole process, for example, while checking flaky specs.

**Comment by filipefurtad0** [comment]
Following @sigmundpetersen advice, all the jobs/nodes are running from the one build file only :+1: 

I've taken this additional round of changes to remove the unused and deprecated `knapsack gem`. 

Moving again to code review :+1:

**Comment by dacook** [comment]
> * build runs should be visible under the project's dashboard

@filipefurtad0 are you referring to a Knapsack Pro dashboard, or just the GitHub Actions screen like below?

Screen Shot 2022-11-22 at 9 37 08 am

----

And wow, I don't believe my eyes!!! 🤩 

> ### Total Duration 3m 19s

**Review by sigmundpetersen — APPROVED** [review]

**Comment by jibees** [comment]
`3m19s`

😊 🤯 💯 🚴‍♂️ 🚴‍♀️

**Comment by filipefurtad0** [comment]
Hey @dacook ,

> @filipefurtad0 are you referring to a Knapsack Pro dashboard

Yes, I mean the Knapsack dashboard. I'll add you to the team so you can as well check status on specs and run times :+1: 

> it only removes duplication of some lines, and one large file could be harder to navigate than the smaller ones. 

True - the size of the build file is larger indeed, I hope we don't need to change it often.

The main advantage of having all jobs in one file is to be able to access these on Github Actions under the same URL (the screen you've [posted before](https://github.com/openfoodfoundation/openfoodnetwork/pull/9895#issuecomment-1322750293)). Before, each job had a separate URL in GA, so one would have to click around a lot, to track failing tests.

> Total Duration 3m 19s

I think this was only for the last run of a failing job :-) The slowest job seems to run in 5 mins or so, but yeah, still an improvement :sweat_smile:

**Comment by dacook** [comment]
> The main advantage of having all jobs in one file is to be able to access these on Github Actions under the same URL

Ah ha, thanks for pointing this out.

I created an account with Knapsack Pro, so hopefully you can add my account to the project now.

Do you think we should wait for a re-review from Maikel?
Otherwise, if all looks good on the Knapsack Pro side, then please go ahead and merge!!

**Comment by jibees** [comment]
> then please go ahead and merge!!

👍

**Comment by dacook** [comment]
Hi @filipefurtad0 , I can't see how the Knapsack dashboard relates to this setup. I see test suite names like "system_specs" and "controller tests", but these aren't the same as the job names, for example "knapsack_rspec_controllers".
Can you please help to explain what we're seeing in the Knapsack dashboard?

[Screen Shot 2022-11-24 at 10 17 30 am](https://user-images.githubusercontent.com/4188088/203662447-d8e21bc3-ed67-490c-994e-c3d47765c8d8.png)

**Comment by dacook** [comment]
A spec failure, but I guess it is a flaky system spec:
https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/3537901902/jobs/5938277256

The next master build succeeded 😌 
https://github.com/openfoodfoundation/openfoodnetwork/actions/workflows/build.yml?query=branch%3Amaster

**Comment by filipefurtad0** [comment]
> these aren't the same as the job names, for example "knapsack_rspec_controllers"

Good point @dacook ,
I've updated the test suite names on the Knapsack pro dashboard, to match the jobs on `build.yml`:

image

While doing so, I noticed that engines specs were using the same API key as models specs, which is not good practice. So I've created #10053 to address this - which in turn surfaced some syntax issues, leading to missing folders (all explained in the PR).

**Comment by mkllnk** [comment]
Oh my god, this is amazing. Running the whole build in a few minutes? :exploding_head:

**Comment by filipefurtad0** [comment]
> Oh my god, this is amazing. Running the whole build in a few minutes? exploding_head

As a FYI @mkllnk, there are a few ongoing issues though, which occur from time to time: 
- https://github.com/openfoodfoundation/openfoodnetwork/issues/10018 - passes on second run
- https://github.com/openfoodfoundation/openfoodnetwork/issues/10220 - passes on second run
- Order related spec failures: currently, we have [queue mode](https://knapsackpro.com/faq/question/why-when-i-use-queue-mode-for-rspec-then-my-tests-fail) enabled for system specs. This speeds up the build but can lead to order related failures. This is one example #10055.

My perception is, these don't happen often, and even when, usually re-triggering the failing node costs us a few minutes only. It's not ideal, but I have not made much (or any!) progress on the issues above. 

Please let me know if you have a different perception on this, or any suggestions for improvement. Thanks!!

**Comment by mkllnk** [comment]
We should be able to fix the order dependent runs. Browser crashing is probably more difficult to pin down. But it could happen with the old build already, maybe less often. And rebuilding the old build took a lot longer. This new system with Knapsack Pro is excellent.

**Pull request merged**
